File: src/main/java/frc/robot/subsystems/vision/CameraIO.java
package frc.robot.subsystems.vision;

import org.littletonrobotics.junction.AutoLog;

import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Transform3d;
import edu.wpi.first.units.measure.Distance;

public interface CameraIO {
    @AutoLog
    public static class CameraIOInputs {
        public boolean connected = false;
        public PoseObservation[] poseObservations = new PoseObservation[0];
        public int[] tagIds = new int[0];
        public TargetObservation latestTargetObservation = new TargetObservation(new Rotation2d(), new Rotation2d(),
                -1);
        public TargetObservation[] targets = new TargetObservation[0];

        public Transform3d robotToCamera = new Transform3d();
        public String name = "";
    }

    /** Represents the angle to a simple target, not used for pose estimation. */
    public static record TargetObservation(Rotation2d tx, Rotation2d ty, int ID) {
    }

    /** Represents a robot pose sample used for pose estimation. */
    public static record PoseObservation(
            double timestamp,
            Pose3d pose,
            double ambiguity,
            int tagCount,
            Distance averageTagDistance) {
    }

    public default void updateInputs(CameraIOInputs inputs) {
    }
}

File: src/main/kotlin/frc/robot/subsystems/vision/CameraIO.kt
package frc.robot.subsystems.vision

import edu.wpi.first.math.geometry.Pose3d
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.geometry.Transform3d
import org.littletonrobotics.junction.AutoLog

interface CameraIO {
    fun updateInputs(inputs: CameraIOInputs) {}

    @AutoLog
    open class CameraIOInputs {
        @JvmField var connected: Boolean = false
        @JvmField var poseObservations: Array<PoseObservation> = emptyArray()
        @JvmField var tagIds: IntArray = IntArray(0)
        @JvmField var latestTargetObservation: TargetObservation = TargetObservation(Rotation2d(), Rotation2d(), -1)
        @JvmField var targets: Array<TargetObservation> = emptyArray()
        @JvmField var robotToCamera: Transform3d = Transform3d()
        @JvmField var name: String = ""
    }
}
